pkg/storage/storage.go

```diff
@@ -40,14 +41,11 @@ func NewStorage() *storage {
 
 // Ready returns true if metrics-server's storage has accumulated enough metric
 // points to serve NodeMetrics.
-// Checking only nodes for metrics-server's storage readiness is enough to make
-// sure that it has accumulated enough metrics to serve both NodeMetrics and
-// PodMetrics. It also covers cases where metrics-server only has to serve
-// NodeMetrics.
 func (s *storage) Ready() bool {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
-	return len(s.nodes.prev) != 0
+	// Avoid marking ready after one scrape with restarted container
+	return s.storeIndex >= 2 && len(s.nodes.prev) != 0 || len(s.pods.prev) != 0
```

Comment:
PTAL https://github.com/kubernetes-sigs/metrics-server/pull/766

Reply:
With this change this will no longer be true

Reply:
ah indeed, I missed that. But still, we shouldn't need a storeIndex. `len(s.nodes.prev) != 0 || len(s.pods.prev) != 0` should be enough to be able to compute the CPU usage of either the pods or the nodes so even with only 1 scrape the storage could theoretically be ready.

Reply:
For ready I mostly think about how rolling upgrade should work. It should be smooth experience, where new instance takes over when it has similar data to the previous one. With https://github.com/kubernetes-sigs/metrics-server/pull/766 MS can detect fresh containers or container that have restarted and use it's container start time point to create both point in `last` and `prev` map in one scrape. In case of rolling update, new instance could pick up a freshly started container and mark itself as ready, even though in reality it can serve metrics from only 1 container vs whole cluster.

Reply:
I'm still uncertain with this change, even though the computation for pods was improved to only require 1 scrape, we would still make MS unready even though it can already serve PodMetrics. In the end the store index here only use is to leave some time for node metrics to be collected.
Your initial reasoning for checking both nodes and pods was to improve stability, but by adding the store index, we still expect node information even if they might not be here. At this point it comes down to whether MS should be ready once it can return metrics or once it can return metrics for either pods or nodes and theoretically for both of them (the new change). As you might be able to tell, I'm still having a hard time with the storeIndex which is a very theoretical information and that don't tell anything about the data stored.
To me, MS should be ready once it can report any metrics. If we want to make it ready once it can report metrics for both pods and nodes, then we would stumble upon the same instability issues that you've mentioned before.

Reply:
Ok, maybe let's look at this from other side. Becoming ready not nesesery means that new replica will start serving traffic during rolling upgrade. I have added `initialDelaySeconds` to the readiness probe. It means that MS should not testing readiness or sending traffic within first 20 seconds achieving the same result. It may be performance sensitive, but I think it's sensible compromise for now.

Reply:
SG